doc/oauthsign.1
.\"                              hey, Emacs:   -*- nroff -*-
.\" oauth-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH OAUTHSIGN 1 "October 16, 2008"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.\"
.\" defines for verbose text
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..

.SH NAME
oauthsign \- oauth command line tool
.SH SYNOPSIS
.B oauthsign
.RI [ options ]
.RI < URL >
.RI [ CKey ]
.RI [ CSec ]
.RI [ TKey ]
.RI [ Tsec ]
.SH DESCRIPTION
\fBoauthsign\fP offers a simple way to generate or make OAuth signed HTTP requests.
.PP
\fBoauthsign\fP implements an OAuth consumer...
.SH OPTIONS
\fBoauthsign\fP accepts the following options:
.TP
.B  --dry-run
Take no real actions.
.TP
.B  --no-warn
Disable warnings.
.TP
.BR  -q , " --quiet" , " --silent"
Inhibit usual output.
.TP
.B  -v, --verbose
Print more information.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-V, \-\-version
Show version of program.
.TP
.B \-b, \-\-base-string
Print OAuth base-string and exit.
.TP
.B \-B, \-\-base-url
Print OAuth base-URL and exit.
.TP
.B \-r, \-\-request <type>
Set the HTTP request type (HEAD, PUT, POST, GET [default],..).
.TP
.B \-p, \-\-post
Same as '-r POST'
.TP
.B \-d, \-\-data <key>[=<value>]
Add key=value URL query parameters. Both key and value are expected to be plain ASCII. oauthsign will URL-encode them.
.TP
.B \-m, \-\-signature-method <method>
Specify the oauth signature method (PLAINTEXT, RSA-SHA1, HMAC-SHA1 [default]).
.TP
.B \-c, \-\-CK, \-\-consumer-key <text>
Set consumer-key.
.TP
.B \-C, \-\-CS, \-\-consumer-secret <text>
Set consumer-secret.
.TP
.B \-t, \-\-TK, \-\-token-key <text>
Set token-key.
.TP
.B \-T, \-\-TS, \-\-token-secret <text>
Set token-secret.
.TP
.B \-f, \-\-file <filename>
Read tokens and secrets from config-file.
.TP
.B \-w
Write tokens to config-file.
.TP
.B \-F <filename>
Set config-file name w/o reading the file. useful for writing to a different file than reading eg. '-f file1 -F file2 -w'
.TP
.B \-x
Make HTTP request and return the replied content.
.TP
.B \-X
Make HTTP request and parse the reply for tokens.
.TP
.B \-\-dry-run
Take no real actions (with -x, -w or -X)
.TP
.B \-e, \-\-erase-token
Clear the [access|request] token (key&secret).
.TP
.B \-E, \-\-erase-all
Wipe consumer and token information and reset signature method to HMAC-SHA1.
.TP
.B \-\-erase-consumer-key
Unset the required consumer-key (don't require a specific consumer-key for the validation to succeed)
.TP
.B \-\-erase-consumer-secret
Unset the consumer-secret
.TP
.B \-\-erase-consumer-key
Unset the required token-key (any token will validate)
.TP
.B \-\-erase-token-secret
Unset the token-secret
.TP
.B \-\-G
Print the request-parameters as url-encoded string according to RFC3986, even if the POST method was choosen.
.TP
.B \-\-P
Toggle parameter escaping for output. with -P, POST params will be escaped and GET parameter won't. This is non-standard behaviour.
.TP
.B \-\-curl
Output a curl commandline. (not with -x or -X)
.PP
The position of parameters -d, -f, -F, -e, -E and all tokens matters!
Tokens are read from file at the moment the -f option is parsed overriding
the current value(s). Optional trailing key/secret params are parsed last.
eg. '\-f config.txt -e -C secret -F out.txt -w' reads the settings from file,
then deletes the access/request tokens and finally overrides the consumer-
secret. Only the consumer-key is left from config.txt and will be saved 
to out.txt along with the new secret. If -X is given and the HTTP request
succeeds, the received token and secret will be stored in out.txt as well.
.PP
The request URL is constructed by first parsing all query-parameters from
the URL; then -d parameters are added, and finally oauth_XYZ params 
appended. The order of the request parameters (while different in the oauth
base-url) is preserved.
.PP
Note that there is also a difference between empty ("") and unset (NULL) parameters: 
an empty parameter is still included in the request, while an unset parameter is not.
use '-e' or '-E' to unset previously set parameters. 
.SH EXAMPLE
run the following commands in sequence to walk though an
.URL "http://term.ie/oauth/example/" "OAuth example".
Connecting to the server requires online connectivity. The file \fI/tmp/test.oaf\fP is used to store the OAuth tokens between the requests.

.PP
.Vb 1
\&oauthsign \-X \-f /tmp/test.oaf \-w \-e \-c key \-C secret http://term.ie/oauth/example/request_token.php
\&oauthsign \-X \-f /tmp/test.oaf \-w http://term.ie/oauth/example/access_token.php
\&oauthsign \-x \-f /tmp/test.oaf "http://term.ie/oauth/example/echo_api.php?method=foo%20bar&bar=baz"
\&oauthsign \-x \-f /tmp/test.oaf \-d method=foo%20bar \-d "bar=baz &test%" \-\-post http://term.ie/oauth/example/echo_api.php
.Ve

.SH "SEE ALSO"
.BR oauth-utils (7).
.SH AUTHOR
oauth-utils was written by Robin Gareus <robin@gareus.org>.
.PP
This manual page was also written by Robin Gareus <robin@gareus.org>,
for the Debian project (but may be used by others).



ChangeLog
2009-08-17  Robin Gareus  <robin@gareus.org>
	* version 0.3.1
	* minor fixes ('-p' parameter)

2009-07-31  Robin Gareus  <robin@gareus.org>
	* version 0.3.0
	* added support for GET/PUT/HEAD request (Stephen Paul Weber)
	* commandline options for OAuth Code 1.0 Rev A

2008-10-19  Robin Gareus  <robin@gareus.org>
	* version 0.2.3
	* added oauthverify
	* updated man-pages
	* reworked output (stdout/stderr)
	* enforce required oauth parameters

2008-10-17  Robin Gareus  <robin@gareus.org>
	* version 0.2.0
	* escape parameters when saving keyfile
	* added curl commandline output
	* reading SHA1-key from consumer-secret :(

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.5
	* fixed segfault on filename-check.

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.4
	* updated debian packaging system

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.3
	* renamed oauth_utils to oauth-utils
	* debianized it
	* fixed typos, built-env and updated example.sh
	* changed parser to allow than two return parameters (xoauth_)

2008-10-16  Robin Gareus  <robin@gareus.org>
	* version 0.1.0
	* oauthsign works and is mostly documented.

2008-10-13  Robin Gareus  <robin@gareus.org>

	* oauth_utils: initial version.



src/oauthsign.c
/* 
   oauthsign - command line oauth

   Copyright (C) 2008, 2009 Robin Gareus

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  

#include <termios.h>
#include <grp.h>
#include <pwd.h>
*/

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdio.h>
#include <sys/types.h>
#include <getopt.h>
#include "system.h"

#include "oauth_common.h"

#define EXIT_FAILURE 1

char *xmalloc ();
char *xrealloc ();
char *xstrdup ();

static void usage (int status);

/* The name the program was run with, stripped of any leading path. */
char *program_name;

/* getopt_long return codes */
enum {DUMMY_CODE=129
     ,DRYRUN_CODE
     ,NOWARN_CODE
     ,CURLOUT_CODE
     ,NULLCK_CODE
     ,NULLCS_CODE
     ,NULLTK_CODE
     ,NULLTS_CODE
};

/* Option flags and variables */

int no_warnings  = 0; /* --no-warn */
int want_quiet   = 0; /* --quiet, --silent */
int want_verbose = 0; /* --verbose */
int want_dry_run = 0; /* --dry-run */

int mode         = 1; ///< mode: 1=GET 2=POST; general operation-mode - bit coded 
                      //  bit0 (1)  : enable ?! (also see want_dry_run)
                      //  bit1 (2)  : HTTP POST enable (no GET) 
                      //  bit2 (4)  : (unused ; old oauthrequest() compat)
                      //  bit3 (8)  : -b base-string and exit
                      //  bit4 (16) : -B base-url and exit
                      //  bit5 (32) :  print secrets along with base-string(8)
                      //  bit6 (64) :  print signature after generating it. (unless want_verbose is set: it's printed anyway)
                      //  bit7 (128):  parse reply (request token, access token)
                      //  bit8 (256):  toggle ouput Parameter escape. (dont escape GETs and escape POST params with format_array(..)
                      //  bit9 (512):  curl-output
                      //
char *method;
int request_mode = 0; ///< mode: 0=print info only; 1:perform HTTP request

int print_as_get = 0; 
int want_write   = 0;
int   oauth_argc = 0;
char **oauth_argv = NULL;
char *datafile   = NULL;
char *callback = NULL;
char *verifier = NULL;
oauthparam op;

static struct option const long_options[] =
{
  {"quiet", no_argument, 0, 'q'},
  {"silent", no_argument, 0, 'q'},
  {"verbose", no_argument, 0, 'v'},
  {"dry-run", no_argument, 0, DRYRUN_CODE},
  {"no-warn", no_argument, 0, NOWARN_CODE},
  {"help", no_argument, 0, 'h'},
  {"version", no_argument, 0, 'V'},

  {"consumer-key", required_argument, 0, 'c'},
  {"consumer-secret", required_argument, 0, 'C'},
  {"token-key", required_argument, 0, 't'},
  {"token-secret", required_argument, 0, 'T'},
  {"CK", required_argument, 0, 'c'},
  {"CS", required_argument, 0, 'C'},
  {"TK", required_argument, 0, 't'},
  {"TS", required_argument, 0, 'T'},
  {"callback", required_argument, 0, 'a'},
  {"verifier", required_argument, 0, 'A'},
  {"erase-consumer-key", no_argument, 0, NULLCK_CODE},
  {"erase-consumer-secret", no_argument, 0, NULLCS_CODE},
  {"erase-token-key", no_argument, 0, NULLTK_CODE},
  {"erase-token-secret", no_argument, 0, NULLTS_CODE},
  {"signature-method", required_argument, 0, 'm'}, //  oauth signature method

  {"request", required_argument, 0, 'r'}, // HTTP request method (GET, POST)
  {"post", no_argument, 0, 'p'},
//{"toggle-escape", no_argument, 0, 'P'}, // TODO: rename ''--print-escaped'' ?!
  {"curl", no_argument, 0, CURLOUT_CODE}, 
  {"data", required_argument, 0, 'd'},
  {"base-url", no_argument, 0, 'B'},
  {"base-string", no_argument, 0, 'b'},

  {"file", required_argument, 0, 'f'},
  {"write", no_argument, 0, 'w'}, 
//{"writefile", required_argument, 0, 'F'}, 
//{"execute", no_argument, 0, 'x'}, 
//{"oauthrequest", no_argument, 0, 'X'}, 
  {NULL, 0, NULL, 0}
};

/** Set all the option flags according to the switches specified.
 *  Return the index of the first non-option argument.  
 */
static int decode_switches (int argc, char **argv) {
  int c;

  while ((c = getopt_long (argc, argv, 
			   "h"	/* help */
			   "V" 	/* version */
			   "q"	/* quiet or silent */
			   "v"	/* verbose */
			   "b"  /* base-string*/
			   "B"  /* base-URL*/
			   "r:" /* HTTP Request method */
			   "p" 	/* post */
			   "d:" /* URL-query parameter data eg.
               * '-d name=daniel -d skill=lousy'->'name=daniel&skill=lousy' */
			   "m:" /* oauth signature Method */
			   "P" 	/* print escaped post parameters */
			   "G" 	/* always print as GET parameters */

			   "c:" /* consumer-key*/
			   "C:" /* consumer-secret */
			   "t:" /* token-key*/
			   "T:" /* token-secret */

			   "f:" /* read key/data file */
			   "w" 	/* write to key/data file, save request/access token state */
			   "F:" /* set key/data filename */
			   "x" 	/* execute */
			   "X" 	/* execute and parse reply */
			   "e"  /* erase token */
			   "E", /* erase token and conusmer */
			   long_options, (int *) 0)) != EOF) {
    switch (c) {
      case 'q':		/* --quiet, --silent */
        want_quiet = 1;
        break;
      case 'v':		/* --verbose */
        want_verbose = 1;
        break;
      case NOWARN_CODE:	/* --no-warn */
        no_warnings = 1;
        break;
      case DRYRUN_CODE:	/* --dry-run */
        want_dry_run = 1;
        break;
      case CURLOUT_CODE:	/* --curl */
        mode|= 512;
        break;
      case 'V':
        printf ("%s %s (%s)", PACKAGE, VERSION, OS);
        #ifdef LIBOAUTH_VERSION
        printf (" liboauth/%s", LIBOAUTH_VERSION);
        #endif
        printf ("\n");
        exit (0);

      case 'b':
        mode&=~(8|16);
        mode|=8;
        break;
      case 'B':
        mode&=~(8|16);
        mode|=16;
        break;
      case 'p':
        mode&=~(1|2);
        mode|=2;
        free(method); method=xstrdup("POST");
        break;
      case 'r':
        mode&=~(1|2|4);
        if (!strncasecmp(optarg,"GET",3)) {
          mode|=1;
          free(method); method=xstrdup("GET");
     // else if (!strncasecmp(optarg,"POSTREQUEST",4))
     //   mode|=4;
        } else if (!strncasecmp(optarg,"POST",4)) {
          mode|=2;
          free(method); method=xstrdup("POST");
        } else {
          int i;
          mode|=1; // XXX
          method = strdup(optarg); 
          for (i=0;i<strlen(method);i++) method[i]=toupper(method[i]);
        }
        break;
      case 't':
        if (op.t_key) free(op.t_key);
        op.t_key=xstrdup(optarg); 
        break;
      case 'T':
        if (op.t_secret) free(op.t_secret);
        op.t_secret=xstrdup(optarg); 
        break;
      case 'a':
#if 1
        if (strlen(optarg) == 0) 
          callback=xstrdup("oob");
        else
#endif
        callback=xstrdup(optarg); 
        break;
      case 'A':
        verifier=xstrdup(optarg); 
        break;
      case 'c':
        if (op.c_key) free(op.c_key);
        op.c_key=xstrdup(optarg); 
        break;
      case 'C':
        if (op.c_secret) free(op.c_secret);
        op.c_secret=xstrdup(optarg); 
        break;
      case 'd': 
        add_param_to_array(&oauth_argc, &oauth_argv,optarg);
        break;
      case 'f':
        read_keyfile(optarg, &op);
      case 'F':
        if (datafile) free(datafile);
        datafile=xstrdup(optarg);
        break;
      case 'w':
        want_write=1;
        break;
      case 'X':
        mode|=128; // parse reply and enter request mode..
      case 'x':
        request_mode=1;
        break;
      case 'e':
        reset_oauth_token(&op);
        break;
      case 'E':
        reset_oauth_param(&op);
        break;
      case NULLCK_CODE:	/* --erase-consumer-key */
        if (op.c_key) free(op.c_key);
        op.c_key=NULL;
        break;
      case NULLCS_CODE:	/* --erase-consumer-secret */
        if (op.c_secret) free(op.c_secret);
        op.c_secret=NULL;
        break;
      case NULLTK_CODE:	/* --erase-token-key */
        if (op.t_key) free(op.t_key);
        op.t_key=NULL;
        break;
      case NULLTS_CODE:	/* --erase-token-secret */
        if (op.t_secret) free(op.t_secret);
        op.t_secret=NULL;
        break;
      case 'G':
        print_as_get=1;
        break;
      case 'P':
        mode|=256;
        break;
      case 'm':
        if (parse_oauth_method(&op, optarg)) usage(1);
        break;
      case 'h':
        usage (0);

      default:
        usage (EXIT_FAILURE);
    }
  }
  return optind;
}


static void
usage (int status)
{
  printf (_("%s - \
command line utilities for oauth\n"), program_name);
  printf (_("Usage: %s [OPTION]... URL [CKey] [CSec] [TKey] [Tsec]\n"), program_name);
  printf (_("\
Options:\n\
  -h, --help                  display this help and exit\n\
  -V, --version               output version information and exit\n\
  -q, --quiet, --silent       inhibit usual output\n\
  -v, --verbose               print more information\n\
  --no-warn                   dont print any warnings.\n\
  \n\
  -b, --base-string           print OAuth base-string and exit\n\
  -B, --base-url              print OAuth base-URL and exit\n\
"/*
  --curl                      format output as `curl` commandline\n\
*/"\
  -r, --request <type>        HTTP request type (HEAD, POST, GET [default],..)\n\
  -p, --post                  same as -r POST\n\
  -d, --data <key>[=<val>]    add ASCII url-query parameters.\n\
                              Note: oauthsign will URL-encode key & value.\n\
  -m, --signature-method <m>  oauth signature method (PLAINTEXT,\n\
                              RSA-SHA1, HMAC-SHA1 [default])\n\
"/*
  -P,                         print URL-escaped POST parameters\n\ 
*/"\
  \n\
  -c, --CK, --consumer-key    <text> \n\
  -C, --CS, --consumer-secret <text> \n\
  -t, --TK, --token-key       <text> \n\
  -T, --TS, --token-secret    <text> \n\
  \n\
  -a, --callback <url>        specify oauth_callback url (or \'oob') // 1.0 Rev A\n\
  -A, --verifier <text>       specify oauth_verifier // 1.0 Rev A\n\
  \n\
  -f, --file <filename>       read tokens and secrets from config-file\n\
  -w                          write tokens to config-file\n\
  -F <filename>               set config-file name w/o reading the file.\n\
  -x                          make HTTP request and return the replied content\n\
  -X                          make HTTP request and parse the reply for tokens\n\
                              use '-X -w' to request and store tokens.\n\
  --dry-run                   take no real actions (with -x, -w or -X)\n\
  -e, --erase-tokens          clear [access|request] tokens.\n\
  -E, --erase-all             wipe all tokens and reset method to HMAC-SHA1.\n\
  --erase-consumer-key        unset consumer-key\n\
  --erase-consumer-secret     unset consumer-secret\n\
  --erase-token-key           unset token-key\n\
  --erase-token-secret        unset token-secret\n\
  \n\
  The position of parameters -d, -f, -F, -e, -E and all tokens matters!\n\
  \n\
  Tokens are read from file at the moment the -f option is parsed overriding \n\
  the current value(s). Optional trailing key/secret params are parsed last.\n\
  eg.\n\
    '-f config.txt -e -C secret -F out.txt -w' reads the settings from file,\n\
  then deletes the access/request tokens and finally overrides the consumer-\n\
  secret. Only the consumer-key is left from config.txt and will be saved \n\
  to out.txt along with the new secret. If -X is given and the HTTP request\n\
  succeeds, the received token and secret will be stored as well.\n\
  \n\
  The request URL is constructed by first parsing all query-parameters from\n\
  the URL; then -d parameters are added, and finally oauth_XYZ params \n\
  appended.\n\
  \n\
"));
  exit (status);
}


int main (int argc, char **argv) {
  int i;
  int exitval=0;
  int oaargc =0;
  char **oaargv= NULL;
  char *sign = NULL;

  // initialize 

  program_name = argv[0];
  memset(&op,0,sizeof(oauthparam));
  reset_oauth_param(&op);
  method=xstrdup("GET");

  // parse command line

  i = decode_switches (argc, argv);

  if (i>=argc) usage(1);
  op.url=xstrdup(argv[i++]);

  if (argc>i) { if (op.c_key) free(op.c_key); op.c_key=xstrdup(argv[i++]); }
  if (argc>i) { if (op.c_secret) free(op.c_secret); op.c_secret=xstrdup(argv[i++]); }
  if (argc>i) { if (op.t_key) free(op.c_key); op.t_key=xstrdup(argv[i++]); }
  if (argc>i) { if (op.t_secret) free(op.t_secret); op.t_secret=xstrdup(argv[i++]); }
  if (argc>i) 
      usage(EXIT_FAILURE);

  // check settings 

  if (!op.c_key || strlen(op.c_key)<1) {
    fprintf(stderr, "ERROR: consumer key not set.\n");
    exit(1);
  }

  if (want_write && !datafile || (datafile && strlen(datafile)<1)) {
    want_write=0;
    fprintf(stderr, "ERROR: no filename given. use -F or -f.\n");
    exit(1);
  }

  if (!(mode&2) && print_as_get) {
    if (!no_warnings)
      fprintf(stderr, "WARNING: -G is redundant with GET requests.\n");
  } else if (!want_quiet && print_as_get) {
    if (!no_warnings)
      fprintf(stderr, "WARNING: non standard output format.\n");
  }
  if ((mode&256) && !want_quiet) {
    if (!no_warnings)
      fprintf(stderr, "WARNING: non standard parameter escape settings.\n");
  }

  if (want_write && !want_dry_run) { // save current state
    if (save_keyfile(datafile, &op)) {
      want_write=0;
      if (!no_warnings)
        fprintf(stderr, "WARNING: saving state to file '%s' failed.\n", datafile);
    } else if (want_verbose) {
      fprintf(stderr, "saved state to %s\n", datafile);
    }
  }

  if (callback) {
    add_kv_to_array(&oauth_argc, &oauth_argv, "oauth_callback", callback);
  }
 
  if (verifier) { // OAuth 1.0 Rev A
    add_kv_to_array(&oauth_argc, &oauth_argv, "oauth_verifier", verifier);
    free(verifier);
  }
  
  sign = oauthsign_ext(mode, method, &op, oauth_argc, oauth_argv, &oaargc, &oaargv);

  if (sign && want_verbose) 
    fprintf(stderr, "oauth_signature=%s\n", sign);

  if(!request_mode) {
    if (sign) { 
      add_kv_to_array(&oaargc, &oaargv, "oauth_signature", sign);
      free(sign);
    }
    if (print_as_get) mode&=~2; // print as GET !
    format_array(mode, oaargc, oaargv);
  } else { // request_mode 
    char *reply;
    if (!sign) { 
      if (!no_warnings && !want_quiet) 
        fprintf(stderr,"WARNING: could not generate oAuth signature.\n");
      exitval|=8;
    }

    if (!want_dry_run) {
      reply = oauthrequest_ext(mode, &op, oaargc, oaargv, sign);
    } else { 
      if (!want_quiet || want_verbose) 
        fprintf(stderr, "DRY-RUN. not making any HTTP request.\n"); 
      if (sign) add_kv_to_array(&oaargc, &oaargv, "oauth_signature", sign);
      if (print_as_get) mode&=~2; // print as GET !
      format_array(mode, oaargc, oaargv);
      reply=NULL;
    }

    if (!reply) { 
      if (!exitval || want_verbose) fprintf(stderr,"ERROR: no reply from HTTP request.\n");
      exitval|=2;
    } else if (want_verbose || (mode&128)==0) {
      if(want_verbose) fprintf(stderr, "------HTTP reply------\n");
      fprintf(want_verbose?stderr:stdout, "%s\n", reply);
      if(want_verbose) fprintf(stderr, "----------------------\n");
    }

    if (!want_dry_run && (mode&128)) {
      reset_oauth_token(&op);
      int reply_flags;
      if (parse_reply(reply, &(op.t_key), &(op.t_secret), &reply_flags)) { 
        if (!exitval || want_verbose) fprintf(stderr,"ERROR: could not parse reply.\n");
        exitval|=4;
      } else if (!want_quiet) {
        printf ("token=%s\n",op.t_key);
        printf ("token_secret=%s\n",op.t_secret);
        if((reply_flags&1) == 1 && callback) 
          printf ("oauth_callback_confirmed=true # -> OAuth V 1.0 Rev A\n");
      }
    }

    if (sign) free(sign);
    if (reply) free(reply);
  }

  if (callback) {
    // TODO: check if SP returned "oauth_callback_confirmed=true"  -> 1.0 rev A
    free(callback);
  }
 
  if (exitval==0 && want_write && !want_dry_run) { // save final state
    if (save_keyfile(datafile, &op)) {
      want_write=0;
      if (!no_warnings)
        fprintf(stderr, "WARNING: saving state to file '%s' failed.\n", datafile);
    } else if (want_verbose) {
      fprintf(stderr, "saved state to %s\n", datafile);
    }
  }
 
  free_array(oaargc, oaargv);
  free_array(oauth_argc, oauth_argv);
  free(method);
  return (exitval);
}

/* vim: set sw=2 ts=2 sts=2 et : */
